lib/drivers/aws/dedicated_pool.go

```diff
@@ -0,0 +1,573 @@
+				log.Errorf("AWS: dedicated %q: Unable to run scrubbing for host %q: %v", w.name, host_id, err)
+				continue
+			}
+
+			// Removing the host from the list
+			delete(w.active_hosts, host_id)
+		}
+	}
+}
+
+func isHostMac(host *ec2_types.Host) bool {
+	return host.HostProperties != nil && awsInstTypeAny(aws.ToString(host.HostProperties.InstanceType), "mac")
+}
+
+func isMacTooOld(host *ec2_types.Host) bool {
+	return aws.ToTime(host.AllocationTime).Before(time.Now().Add(-24 * time.Hour))
+}
+
+// Check if the host is ready to be released - if it's mac then it should be older then 24h
+func isHostReadyForRelease(host *ec2_types.Host) bool {
+	// Host not used - for sure ready for release (or scrubbing in case of young mac)
+	if !isHostUsed(host) {
+		return true
+	}
+
+	// Mac in scrubbing process (pending) can be released but should be older then 24h
+	if host.State == ec2_types.AllocationStatePending && isHostMac(host) && isMacTooOld(host) {
+		return true
+	}
+
+	return false
+}
+
+// Check if the host is used
+func isHostUsed(host *ec2_types.Host) bool {
+	if host.State == HOST_RESERVED || host.Instances != nil && len(host.Instances) > 0 {
+		return true
+	}
+	return false
+}
+
+// Check how much capacity we have on a host
+func getHostCapacity(host *ec2_types.Host) uint {
+	if host.State != ec2_types.AllocationStateAvailable || host.AvailableCapacity == nil {
+		return 0
+	}
+	// TODO: For now supports only single-type dedicated hosts
+	return uint(aws.ToInt32(host.AvailableCapacity.AvailableInstanceCapacity[0].AvailableCapacity))
+}
+
+// Updates the hosts list every 5 minutes
+func (w *dedicatedPoolWorker) updateDedicatedHostsProcess() ([]ec2_types.Host, error) {
+	defer log.Infof("AWS: dedicated %q: Exited updateDedicatedHostsProcess()", w.name)
+
+	// Balancing the regular update delay based on the scrubbing optimization because it needs to
+	// record the time of host state change and only then the timer to scrubbing will start ticking
+	update_delay := 5 * time.Minute // 5 min by default
+	scrubbing_delay := time.Duration(w.record.ScrubbingDelay)
+	if scrubbing_delay != 0 && scrubbing_delay < 10*time.Minute {
+		update_delay = scrubbing_delay / 2
+	}
+
+	for {
+		time.Sleep(30 * time.Second)
+		// We need to keep the request rate budget, so using a delay between regular updates.
+		// If the dedicated hosts are used often, it could wait for a while due to often updates
+		if w.active_hosts_updated.Before(time.Now().Add(-update_delay)) {
+			if err := w.updateDedicatedHosts(); err != nil {
+				log.Warnf("AWS: dedicated %q: Error happened during the regular hosts update, continue with updated on %q: %v", w.active_hosts_updated, err)
+			}
+		}
+	}
+}
+
+// Will list all the allocated dedicated hosts on AWS with desired zone and tag
+func (w *dedicatedPoolWorker) updateDedicatedHosts() error {
+	// Do not update too often
+	if w.active_hosts_updated.After(time.Now().Add(-10 * time.Second)) {
```

Comment:
reading from `active_hosts_updated` should acquire the mutex

Reply:
fenced with mutex as well